src/bson.c

```diff
@@ -1201,6 +1289,195 @@ static bool php_phongo_bson_state_parse_type(zval* options, const char* name, ph
 	return retval;
 } /* }}} */
 
+static void field_path_map_element_set_info(php_phongo_field_path_map_element* element, php_phongo_bson_typemap_types type, zend_class_entry* ce TSRMLS_DC)
+{
+	element->node_type = type;
+	element->node_ce   = ce;
+}
+
+static void map_add_field_path_element(php_phongo_bson_typemap* map, php_phongo_field_path_map_element* element)
+{
+	/* Make sure we have allocated enough */
+	if (map->field_paths.allocated < map->field_paths.size + 1) {
+		map->field_paths.allocated += PHONGO_FIELD_PATH_EXPANSION;
+		map->field_paths.map = erealloc(map->field_paths.map, sizeof(php_phongo_field_path_map_element) * map->field_paths.allocated);
+	}
+
+	map->field_paths.map[map->field_paths.size] = element;
+	map->field_paths.size++;
+}
+
+static php_phongo_field_path_map_element* field_path_map_element_alloc(void)
+{
+	php_phongo_field_path_map_element* tmp = ecalloc(1, sizeof(php_phongo_field_path_map_element));
+
+	tmp->entry = php_phongo_field_path_alloc(true);
```

Comment:
In this case, the element names should always be `estrdup()`-ed?

Reply:
Yes, and they will be through https://github.com/mongodb/mongo-php-driver/pull/793/files#diff-f7d29a680148f52d6601f59ed787f577R139  if ``true`` is used.